docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,439 @@
+# ADR 002: Lite Client
+
+## Context
+
+Here we describe the architecture for a Tendermint lite node in Rust,
+including a core lite client library. While the lite client library provides the
+essential verification logic according to the [english](TODO) and [TLA+](TODO)
+specs, the lite node uses the lite client library to sync its local state to the
+latest state of the blockchain using a list of full nodes. In addition to lite
+nodes running on client machines making rpc queries to full nodes, the core lite
+client library should also be usable by IBC handlers on blockchains receiving
+data from relayers. While this document only covers the case of the lite node,
+it's important to have clear separation of concerns so that the library can be
+reused in the IBC context.
+
+Most of the necessary data types for the lite client already exist (pubkey/signature, blockid,
+vote), with support for serialization/deserialization. The crux of the lite
+client is verifying validator sets by computing their merkle root, and verifying
+commits by checking validator signatures. The particular data structures used by
+Tendermint have considerably more features/functionality, much of which is not
+needed for the lite client protocol - hence it can be abstracted away from
+the core verification logic.
+
+In addition to the core verification logic, the lite node needs a way to
+receive data from full nodes, to detect conflicting information, and to report
+on conflicts. While there are many ways for a full node to provide bad
+information, what we're really looking for is misbehaviour by the validators,
+which is reflected in conflicting commits (ie. commits for different blocks at
+the same height). 
+
+In what follows we outline the following components of the lite client software:
+
+- core lite client verification library and traits
+- how Tendermint data-structures implement the core lite client traits
+- how a lite node requests information from full nodes and detects conflicts
+
+Note that the architecture of IBC is out of scope, suffice it to say that the
+core libraries (ie. the verification library and the implementation of core
+traits by the Tendermint types) will be re-usable by IBC.
+
+## Decision
+
+Below is a schematic of the components of a lite node:
+
+![Lite Client Diagram](assets/adr-002-image.png)
+
+We take up the components of the diagram in term.
+
+### State
+
+The lite node state contains the following:
+
+- current height (H)
+- last header (H-1)
+- current validators (H)
+
+It also includes some configuration, which contains:
+
+- trusting period
+- initial list of full nodes
+- method (sequential or skipping)
+- trust level (if method==skipping)
+
+The node is initialied with a trusted header for some height H-1
+(call this header[H-1]), and a validator set for height H (call this vals[H]).
```

Comment:
It only actually needs to be initialized with the validator set hash, since it can then get the validator data from the network. (Unless this is more difficult in implementation)

Reply:
true enough - we could take it even further and say its just initialized with a height and header hash, and everything is fetched from a full node. that would be the most light weight, but still before anything can be started, we'd need to fetch the header and the validators, and we could call that fully populated state "initialization". so I guess it's a difference between what the user has to pass in for initialization and what state has to be initialized before the protocol can really start